src/metabase/lib/schema/filter.cljc

```diff
@@ -3,15 +3,116 @@
   a boolean expression."
   (:require
    [metabase.lib.schema.common :as common]
+   [metabase.lib.schema.ref :as ref]
    [metabase.util.malli.registry :as mr]))
 
-(mr/def ::=
+(def ^:private eq-comparable
+  ;; avoid circular refs between these namespaces.
+  [:schema [:ref :metabase.lib.schema.expression/equality-comparable]])
+
+(def ^:private orderable
+  ;; avoid circular refs between these namespaces.
+  [:schema [:ref :metabase.lib.schema.expression/orderable]])
+
+(defn- defclause
+  [sch & args]
+  {:pre [(qualified-keyword? sch)]}
+  (mr/def sch
+    [:catn
+     [:clause [:= (keyword (name sch))]]
+     [:options ::common/options]
+     (into [:args] args)]))
+
+(doseq [op [::and ::or]]
+  (defclause op [:repeat {:min 2} [:schema [:ref ::filter]]]))
+
+(defclause ::not
+  [:schema [:ref ::filter]])
+
+(doseq [op [::= ::!=]]
+  (defclause op
+    [:repeat {:min 2} eq-comparable]))
+
+(doseq [op [::< ::<= ::> ::>=]]
+  (defclause op
+    [:cat orderable orderable]))
+
+(defclause ::between
+  [:catn [:field orderable] [:min orderable] [:max orderable]])
+
+;; sugar: a pair of `:between` clauses
+(defclause ::inside
+  [:catn
+   [:lat-field orderable]
+   [:lon-field orderable]
+   [:lat-max   orderable]
+   [:lon-min   orderable]
+   [:lat-min   orderable]
+   [:lon-max   orderable]])
+
+;; [:= ... nil], [:!= ... nil], [:or [:= ... nil] [:= ... ""]], [:and [:!= ... nil] [:!= ... ""]]
+(doseq [op [::is-null ::not-null ::is-empty ::not-empty]]
+  (defclause op
+    ::ref/field))
```

Comment:
Should these be allowed for all/string expressions?

Reply:
I think `:is-null` and `:not-null` are ok for everything, so we should allow it for any expression.

I don't `:is-empty` or `:not-empty` make sense for everything. For example don't think it makes sense to ask whether a DateTime is empty. But any sort of `::string` expression, including `:type/Text` refs, should be ok